### PR TITLE
Added an initial fuzzer together with infrastructure the works with o…

### PR DESCRIPTION
**Pull request by DavidKorczynski**
…ss-fuzz.

Hi guys,
I added an initial integration with OSS-Fuzz, a service that enables continuous fuzzing run by Google. If you are interested in this I can keep pushing more fuzzers up so we can get good coverage of the project. I placed the fuzzers within the tests/internal directory since I am looking to cover this part of the project first (assuming you are on board).

If you are interested, then please provide me with an email where bug reports will be send to!
If you are against having the fuzzers in your repo then they can be kept in the oss-fuzz repo, however, we would prefer to have them in the fluent-bit repo.

David

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [N/A] Example configuration file for the change
- [N/A] Debug log output from testing the change

- [N/A] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

**Documentation**

- [N/A] Documentation required for this feature

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

**Comment by edsiper** [comment]
thanks for this! highly appreciated.

I am adding my comments in the PR

File: tests/internal/fuzzers/flb_json_fuzzer.cpp [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    int flb_pack_json(char*, int, char**, size_t*, int*);
+}
+
+extern "C" int LLVMFuzzerTestOneInput(unsigned char *data, size_t size)
+{
+	// json packer
+	char *out_buf = NULL;
+	size_t out_size;
+	int root_type;
+	flb_pack_json((char*)data, size, &out_buf, &out_size, &root_type);
```
Comment (edsiper) [inline tests/internal/fuzzers/flb_json_fuzzer.cpp:14]: please validate the return value of flb_pack_json

File: tests/internal/fuzzers/flb_json_fuzzer.cpp [hunk 2]
```diff
@@ -0,0 +1,20 @@
+#include <stdint.h>
+#include <stdlib.h>
+
+extern "C" {
```
Comment (edsiper) [inline tests/internal/fuzzers/flb_json_fuzzer.cpp:4]: why the fuzzer is a .cpp and not a .c ? (we prefer c)

File: tests/internal/fuzzers/flb_json_fuzzer.cpp (hunk 1 above)
Comment (DavidKorczynski) [reply tests/internal/fuzzers/flb_json_fuzzer.cpp]: Done!

File: tests/internal/fuzzers/flb_json_fuzzer.cpp (hunk 2 above)
Comment (DavidKorczynski) [reply tests/internal/fuzzers/flb_json_fuzzer.cpp]: Changed to C now!

**Comment by edsiper** [comment]
how this file is compiled and linked together ?

**Comment by DavidKorczynski** [comment]
This is done from within the OSS-Fuzz project, I have set it up here: https://github.com/google/oss-fuzz/pull/3593. Notice the current build script is for the old commit with the .cpp file and I will update it with the new format once you are happy with the fuzzer. For now I think this is the best approach but as we continue the work and write more fuzzers I will be happy to integrate it further into the fluent-bit build infrastructure. It would be nice to do that, however, as my understanding of the codebase increases.

**Comment by PettitWesley** [comment]
@DavidKorczynski Thank you for working on this!!!

This is something we've wanted to add for a while; continued work on this is highly appreciated :)

**Comment by edsiper** [comment]
@PettitWesley looks like you close it instead of squash it 

note: proper commit: tests: internal: fuzzers: ...

**Comment by PettitWesley** [comment]
@edsiper @DavidKorczynski  Sorry, I only meant to comment 😐

**Comment by edsiper** [comment]
:D

File: tests/internal/fuzzers/flb_json_fuzzer.c [hunk 3]
```diff
@@ -0,0 +1,18 @@
+#include <stdint.h>
+#include <stdlib.h>
+
+int flb_pack_json(char*, int, char**, size_t*, int*);
+
+int LLVMFuzzerTestOneInput(unsigned char *data, size_t size)
+{
+	// json packer
```
Comment (edsiper) [inline tests/internal/fuzzers/flb_json_fuzzer.c:8]: use /* ... */

File: tests/internal/fuzzers/flb_json_fuzzer.c [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+	char *out_buf = NULL;
+	size_t out_size;
+	int root_type;
+	int ret = flb_pack_json((char*)data, size, &out_buf, &out_size, &root_type);
+    
+    if (ret == 0)
```
Comment (edsiper) [inline tests/internal/fuzzers/flb_json_fuzzer.c:14]: fix indentation, and add opening ``` {```

File: tests/internal/fuzzers/flb_json_fuzzer.c [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+        free(out_buf);
+
```
Comment (edsiper) [inline tests/internal/fuzzers/flb_json_fuzzer.c:16]: add closing ```}```

Comment (DavidKorczynski) [reply tests/internal/fuzzers/flb_json_fuzzer.c]: Fixed in new commits!

**Comment by edsiper** [comment]
thanks @DavidKorczynski !

The only missing thing to merge this PR is that you sign-off your commits (git commit -s), without that we cannot merge your contribution.

As soon it is signed we will merge it.

**Comment by DavidKorczynski** [comment]
I think this should do the job? 
I need one thing from you though - an email where the bugs will be reported to and that you will use to see the detailed error reports via Googles ui. This email will be public on the OSS repo.

**Comment by edsiper** [comment]
@DavidKorczynski 

we don't have @fluentbit.io email account, would a mailing list work ? or needs to be a real person email account ? (we can use mine on that case)

**Comment by DavidKorczynski** [comment]
In order to see detailed error reports you need to login with the given account, so it will need to be registered with Google at that point. This is in place as some projects are more sensitive than others and bug reports should not be made public. Because of this, the common approach is to have a central account that maintainers use (setup some password known to all) or simply have a list of emails with the maintainers that are interested in the bug reports. It's no problem to have multiple emails attached. For this reason I suggest we start with your email and then whoever wants to be added we can do so (with your consent).

**Comment by edsiper** [comment]
I understand that makes sense. Please use: edsiper at gmail.com , I will check with other maintainers who want's to be included.

**Comment by DavidKorczynski** [comment]
I added your email @edsiper
I assume the Travis errors has nothing to do with my updates?

**Comment by DavidKorczynski** [comment]
Do I have to do anymore @edsiper or are we ready to merge?

**Comment by edsiper** [comment]
@DavidKorczynski merging now, thanks!

Please add the following maintainer to the Google System: wppttt at amazon.com

**Pull request merged**